src/mlpack/bindings/R/CMakeLists.txt

```diff
@@ -177,19 +181,36 @@ if (BUILD_R_BINDINGS)
        DESTINATION
        "${CMAKE_CURRENT_BINARY_DIR}/mlpack/")
 
+  # This file will take care of multiple definition of functions in .cpp files.
+  file(TOUCH
+      "${CMAKE_CURRENT_BINARY_DIR}/mlpack/model.txt")
+
+  # Generate configure file from configure.ac.
   execute_process(COMMAND autoreconf -i
           WORKING_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/mlpack/")
```

Comment:
I have added `FindAutotools.cmake` to find autoreconf before building R-bindings.

Reply:
Nice, this looks good to me.  The only question I still have (and I have no idea of the answer) is, does R ship with `autoreconf`?  Or is it expected to be available on the system?  The only reason I write this comment is, if R *does* come with `autoreconf` in its distribution, then we can try to use that one even if `find_package(Autotools)` doesn't find one on the system.

Reply:
I am not sure about this, that whether R does come with `autoreconf` or not, May be @coatless or @eddelbuettel can enlighten us more about this.

Reply:
From  https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#DOCF68: 

_R_ has embedded autoconf-2.69 from 2012 and automake-1.16.1 from 2018. Generally, these are invoked by _R_ via `R CMD {build,check}`. That said, I'm not sure they will be found on the `PATH` for non-Linux systems.

Reply:
I know `autoreconf` as _external_ and _optional_ command _sometimes_ used in extended use of the `autoconf` series of commands.  I am almost 100% that this is *not* used regularly by R.  @coatless: `autconf` and `automake` are _different_ binaries.  One generally does _not_ need `autoreconf` and the _shipped_ `configure` is generally good enough.

Do we really require `autoreconf` in a CMake build chain? 


```sh
edd@rob:~$ which autoreconf 
/usr/bin/autoreconf
edd@rob:~$ which autoconf
/usr/bin/autoconf
edd@rob:~$ which automake
/usr/bin/automake
edd@rob:~$
```

Reply:
> Do we really require autoreconf in a CMake build chain?

We can perform this task with `autoconf` also, else we can directly provide `configure` file too.
I can perform this either way, as you people suggest.

Reply:
Maybe we are confusing ourselves between two repos:

- MLPACK.  CMake based. So autotools (as far as I know)

- RcppMLPACK.  R package.  Can use autoconf to generate `configure`.  CMake _could_ be used, is less common.   (We can still use CMake "deep inside" to build MLPACK for use by this package.)  `configure` generally top-level.

Reply:
The build process, when we're all done with this project, to produce a completely working CRAN package should be something like this:

```
git clone https://github.com/mlpack/mlpack
cd mlpack
mkdir build
cd build
cmake -DBUILD_R_BINDINGS=ON ../
make R
```

Whenever that finally completes, then you will have an R package that is ready for submission to CRAN (or very close to it!) in `build/src/mlpack/bindings/R/`.  So, during that step, R is actually *not involved* and so we have to only depend on CMake tooling and infrastructure in a platform-independent way (i.e. we could probably not depend on `autoreconf` or `autotools` during this step, unless we really want to make that a build requirement for the R bindings; I'd prefer to avoid that unless we have to).

However I also understand that after a package is available on CRAN, when a user installs the package from CRAN, there may be some extra steps required.  But in that setting, like @eddelbuettel said, we'll have `autoconf` available to generate the `configure` file (and whatever else needs to be done---about which I'm not very knowledgeable!).

So, I would suggest that if we can, we avoid generating `configure` here and instead do that on the "CRAN side".  Basically, all the CMake code that we directly run as a part of `make R` should only be the minimum of what's necessary to upload the package to CRAN.  I would imagine that just providing a `configure.ac` and not actually an `autoconf`-built `configure` script would be sufficient.

Let me know if I've overlooked something.  I only see the mlpack side of the picture clearly. :)

Reply:
Here I am just asking `cmake` to generate `configure` from `configure.ac` and putting it into the right place(i.e. `build/src/mlpack/bindings/R/mlpack`), and further using configure I am building the package. In Both the repository configure is only used for building the package and no cmake is used for building process. We are only using cmake for properly setting up the package folder and too generate .R and .cpp files.

Reply:
This is good, but I think we can make it clearer still

```sh
git clone https://github.com/mlpack/mlpack
cd mlpack
mkdir build
cd build
cmake -DBUILD_R_BINDINGS=ON ../                    # only CMake up until here
make R                                             # the build invoked by 'make R' may use configure
```

Thing is, once we are at the stage _of an R package_ then in order _build the shippable/installable/... tarball one uses `R CMD build` and when the resulting tarball is installed _it may now contain a `configure` script_.

Clearer?

Reply:
@eddelbuettel so the expectation is that before we upload to CRAN, then `configure` is part of the package?  i.e. it's okay to package `configure` but we should not package `configure.ac`?

Reply:
This is going nowhere.  Let us all get out of the way of @Yashwants19 and let him continue with his work here in generating the bindings.

If and when they are done we can talk (in the other repo) about preparing an R package.   You can then simply look at any differences between the two, which will be trivial to brige.   We are making mountains out of molehills here, so I suggest we stop doing that.   Or at least I should stop reading these frantically typed messages...

Reply:
Bear with me please :)---if you can help me understand what happens on the mlpack side and on the R side, then we can answer the original question I asked, which was whether or not we need to have `autotools` available on the mlpack/CMake side.  That was the original point of the thread that I started.  This step in the code that we are talking about happens during the `cmake -DBUILD_R_BINDINGS=ON ../` step.  So really to put the questions in the clearest way that makes a difference for @Yashwants19's work:

1. Do we need to provide a `configure` as part of what gets submitted to CRAN, instead of a `configure.ac`?  If no, then we don't need autotools in the CMake configuration here.  If yes, then we do, and there's a little more to ask:

2. If we have used CMake to find R already, does R happen to come with an instance of autotools that we could use for this step?  (For instance, Julia ships as an entire distribution with various system libraries hidden in, e.g., `/opt/julia/`.  I don't know if R does anything similar.)  If the answer to the question is no, then we have to depend on it being available from the system.  If yes, then we just need to adapt the CMake scripts that search for autotools to also look in the directories where R might have it.

That's all I'm looking to know---I don't think I'm making mountains out of molehills, mostly I am trying to understand what's needed at which steps of the build so that when we *do* put it all together in the end, we're as close as possible to something that we can just submit upstream to CRAN and be done with. :+1:

Reply:
1. You do not require autotools.  
2. autotools is never "required"
3. It can optionally be used by an R package.

I don't have much more time now, but happy to continue on a quick call this eve, or another eve.

Reply:
Sure, that could be helpful.  Let me know what time works for you.  I have a few questions (that I have hopefully made coherent) about how the R package build process works.

Reply:
Thanks @eddelbuettel for the discussion this morning about the details of what's going on under the hood in the R side.  I believe I have a far more correct understanding now.  I can now ask a much clearer question than I was asking before.

What I am seeing is that the existing RcppMLPACK package---and other Rcpp packages I look at---ship with a `configure` script already: https://github.com/Yashwants19/RcppMLPACK/blob/master/configure

That is, when a user downloads that package from CRAN, it then runs the pre-provided `configure` script.  R doesn't use autotools to generate the `configure` from `configure.ac`, or anything like that.  However, that `configure` script I linked to is *far longer* and more complex than the `configure.ac` file.  Thus, it makes sense from the mlpack side to only include `configure.ac`, and then at the time of mlpack's build to produce the R bindings, we can run autotools (whether it is the one provided by R or not) in order to go from `configure.ac` -> `configure`, which we can then distribute as part of the CRAN package.

@Yashwants19 all of this to say, I think it was the right thing to add `FindAutotools.cmake` since we aren't guaranteed to have that at the time we are building mlpack.  Thank you for doing that. :+1:  It would be great if we could also update the search location to try and find the embedded version that R ships too, based on this comment: https://github.com/Yashwants19/mlpack/pull/7#discussion_r449687768

However, I don't think that's necessary to do right now.  Up to you.  (Honestly, most systems that have CMake available will *also* have autotools available, so finding the embedded R version may not ever be necessary in practice.)

Reply:
[ _If_ one uses `autoconf` to generate `configure` from `configure.ac` (as is common) then yes, `configure` becomes large--100kb or more.  But it is also selfcontained, no other binary needed,  And yes, the `configure` will be pre-made by the author / packager and the user generally never ever builds it.  And, just to re-iterate, with R we can have scripts called `configure` that can be anything as long as they have this name, and are executable.  And again, they are optional.  We do not have to have one if we don;t need one. ]

Reply:
I have removed `configure.ac` file.